homeassistant/components/fibaro/__init__.py

```diff
@@ -242,33 +253,44 @@ def _read_devices(self):
                         self.hub_serial, device.id)
                     self._device_map[device.id] = device
                     self.fibaro_devices[device.mapped_type].append(device)
-                else:
-                    _LOGGER.debug("%s (%s, %s) not used",
-                                  device.ha_id, device.type,
-                                  device.baseType)
+                _LOGGER.debug("%s (%s, %s) -> %s. Prop: %s Actions: %s",
+                              device.ha_id, device.type,
+                              device.baseType, device.mapped_type,
+                              str(device.properties), str(device.actions))
             except (KeyError, ValueError):
                 pass
 
 
-def setup(hass, config):
+def setup(hass, base_config):
     """Set up the Fibaro Component."""
-    hass.data[FIBARO_CONTROLLER] = controller = \
-        FibaroController(config[DOMAIN][CONF_USERNAME],
-                         config[DOMAIN][CONF_PASSWORD],
-                         config[DOMAIN][CONF_URL],
-                         config[DOMAIN][CONF_PLUGINS],
-                         config[DOMAIN])
+    config = base_config.get(DOMAIN)
+    gateways = config.get(CONF_GATEWAYS)
+    hass.data[FIBARO_CONTROLLERS] = {}
 
     def stop_fibaro(event):
         """Stop Fibaro Thread."""
         _LOGGER.info("Shutting down Fibaro connection")
-        hass.data[FIBARO_CONTROLLER].disable_state_handler()
-
-    if controller.connect():
-        hass.data[FIBARO_DEVICES] = controller.fibaro_devices
+        for controller in hass.data[FIBARO_CONTROLLERS].values():
+            controller.disable_state_handler()
+        hass.data[FIBARO_CONTROLLERS] = {}
```

Comment:
If we don't need this key anymore we should pop it or delete it from the dict.

Reply:
Wouldn't that be a less efficient algorithm, tho?
I know it's highly theoretical either way, since it's usually a single element list and it's probably equally readable either way, but this seems like a more appropriate solution on an embedded device. I could be wrong, I'm new here. Of course if you insist, I'll modify it.

Reply:
I don't think we should worry about speed here, but clean up remaining references.

Reply:
Ok, I'll keep in mind.

Reply:
Please remove this line.

Reply:
Ok, I removed it. Just for my education, what was wrong with that line? I'm new to python but as far as I understand it did remove the reference to all items in the map and by that freeing up the memory. I understand your suggestion that you'd prefer enumerate + pop, but coming from decades of embedded c programming and knowing how enumerate and pop works, that's really a very wasteful way of solving the problem. Afaik, enumerate iterates through the list an extra time, creates a new index, and pop unlinks items one-by-one from the list. I could be totally wrong on my assumptions, but in general this construct feels like something to discourage on an embedded platform. I could be totally not pythonic in thinking this.
Anyhow, I'd love to learn. Thanks in advance.